### PR TITLE
Optimize type checking for unions with error or nil

### PR DESCRIPTION
**Pull request by riyafa**
Optimize type checking with nil
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/31568
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/31569
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/31742

## Samples
 > N/A

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeTestGen.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+import static org.objectweb.asm.Opcodes.IFNULL;
+import static org.objectweb.asm.Opcodes.INSTANCEOF;
+import static org.objectweb.asm.Opcodes.INVOKESTATIC;
+import static org.wso2.ballerinalang.compiler.bir.codegen.JvmConstants.BERROR;
+import static org.wso2.ballerinalang.compiler.bir.codegen.JvmConstants.OBJECT;
+import static org.wso2.ballerinalang.compiler.bir.codegen.JvmConstants.TYPE;
+import static org.wso2.ballerinalang.compiler.bir.codegen.JvmConstants.TYPE_CHECKER;
+
+/**
+ * BIR function to JVM byte code generation class.
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeTestGen.java:44]: Shall we update the doc comment with more information regarding the type test optimization?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/typechecker/NilOptimizationTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import org.ballerinalang.test.BRunUtil;
+import org.ballerinalang.test.CompileResult;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.nio.file.Paths;
+
+/**
+ * Tests the optimizations done for error related type checks, to not pass through the TypeChecker.
```
Comment (HindujaB) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/typechecker/NilOptimizationTest.java:31]:
```suggestion
 * Tests the optimizations done for nil related type checks, to not pass through the TypeChecker.
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/typechecker/ErrorOptimizationTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Test(dataProvider = "FunctionList")
+    public void testErrorOptimizations(String funcName) {
+        BRunUtil.invoke(compileResult, funcName);
+    }
+
+    @DataProvider(name = "FunctionList")
+    public Object[][] getTestFunctions() {
+        return new Object[][]{
```
Comment (HindujaB) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/typechecker/ErrorOptimizationTest.java:50]: Shall we return a `String[]` as we only pass the function names here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeTestGen.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        types = Types.getInstance(compilerContext);
+        this.mv = mv;
+        this.jvmTypeGen = jvmTypeGen;
+    }
+
+    void generateTypeTestIns(BIRNonTerminator.TypeTest typeTestIns) {
+        var sourceValue = typeTestIns.rhsOp.variableDcl;
+        BType sourceType = sourceValue.type;
+        BType targetType = typeTestIns.type;
+        if (canOptimizeNilCheck(sourceType, targetType) ||
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeTestGen.java:66]: Shall we add comment here canOptimizeNilCheck and canOptimizeErrorCheck.

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**